docs/malcolm-config.md
# <a name="ConfigAndTuning"></a>Malcolm Configuration

Malcolm's runtime settings are stored (with a few exceptions) as environment variables in configuration files ending with a `.env` suffix in the `./config` directory. The `./scripts/configure` script can help users configure and tune these settings. For an in-depth treatment of the configuration script, see the **Configuration** section in [**End-to-end Malcolm and Hedgehog Linux ISO Installation**](malcolm-hedgehog-e2e-iso-install.md#MalcolmConfigItems).

## <a name="MalcolmConfigEnvVars"></a>Environment variable files

Although the configuration script automates many of the following configuration and tuning parameters, some environment variables of particular interest are listed here for reference.

* **`arkime.env`** and **`arkime-secret.env`** - settings for [Arkime](https://arkime.com/)
    - `ARKIME_AUTO_ANALYZE_PCAP_THREADS` – the number of threads available to Arkime for analyzing PCAP files (default `1`)
    - `ARKIME_PASSWORD_SECRET` - the password hash secret for the Arkime viewer cluster (see `passwordSecret` in [Arkime INI Settings](https://arkime.com/settings)) used to secure the connection used when Arkime viewer retrieves a PCAP payload for display in its user interface
    - `ARKIME_ROTATE_INDEX` - how often (based on network traffic timestamp) to [create a new index](https://arkime.com/settings#rotateIndex) in OpenSearch
    - `ARKIME_QUERY_ALL_INDICES` - whether or not Arkime should [query all indices](https://arkime.com/settings#queryAllIndices) instead of trying to calculate which ones pertain to the search time frame (default `false`)
    - `ARKIME_SPI_DATA_MAX_INDICES` - the maximum number of indices for querying SPI data, or set to `-1` to disable any max. The [Arkime documentation](https://arkime.com/settings#spiDataMaxIndices) warns "OpenSearch/Elasticsearch MAY blow up if we ... search too many indices." (default `7`)
    - `MANAGE_PCAP_FILES` and `ARKIME_FREESPACEG` - these variables deal with PCAP [deletion by Arkime](https://arkime.com/faq#pcap-deletion), see [**Managing disk usage**](#DiskUsage) below
    - `MAXMIND_GEOIP_DB_ACCOUNT_ID` and `MAXMIND_GEOIP_DB_LICENSE_KEY` - Malcolm uses MaxMind's free GeoLite2 databases for GeoIP lookups. As of December 30, 2019, these databases are [no longer available](https://blog.maxmind.com/2019/12/18/significant-changes-to-accessing-and-using-geolite2-databases/) for download via a public URL. Instead, they must be downloaded using a MaxMind license key (available without charge [from MaxMind](https://www.maxmind.com/en/geolite2/signup)). The MaxMind acount ID and license key can be specified here for GeoIP database downloads during build- and run-time.
    - `MAXMIND_GEOIP_DB_ALTERNATE_DOWNLOAD_URL` - As an alternative to (or fallback for) `MAXMIND_GEOIP_DB_LICENSE_KEY`, a URL prefix may be specified in this variable (e.g., `https://example.org/foo/bar`) which will be used as a fallback. This URL should serve up `.tar.gz` files in the same format as those provided by the official source (see the example [here](contributing-github-runners.md#secrets-and-variables)).
    - The following variables configure [Arkime's use](index-management.md#ArkimeIndexPolicies) of OpenSearch [Index State Management (ISM)](https://opensearch.org/docs/latest/im-plugin/ism/index/) or Elasticsearch [Index Lifecycle Management (ILM)](https://www.elastic.co/guide/en/elasticsearch/reference/current/index-lifecycle-management.html):
        + `INDEX_MANAGEMENT_ENABLED` - if set to `true`, Malcolm's instance of Arkime will [use these features](https://arkime.com/faq#ilm) when indexing data; note that this only takes effect when initializing Malcolm from an [empty state](running.md#Wipe)
        + `INDEX_MANAGEMENT_OPTIMIZATION_PERIOD` - the period in hours or days that Arkime will keep records in the **hot** state (default `30d`)
        + `INDEX_MANAGEMENT_RETENTION_TIME` - the period in hours or days that Arkime will keep records before deleting them (default `90d`)
        + `INDEX_MANAGEMENT_OLDER_SESSION_REPLICAS` - the number of replicas for older sessions indices (default `0`)
        + `INDEX_MANAGEMENT_HISTORY_RETENTION_WEEKS` - the retention time period (weeks) for Arkime history data (default `13`)
        + `INDEX_MANAGEMENT_SEGMENTS` - the number of segments Arlime will use to optimize sessions (default `1`)
        + `INDEX_MANAGEMENT_HOT_WARM_ENABLED` - whether or not Arkime should use a hot/warm design (storing non-session data in a warm index); setting up hot/warm index policies also requires configuration on the local nodes in accordance with the [Arkime documentation](https://arkime.com/faq#ilm)
    - The following variables configure exposing [Arkime's WISE Plugin](https://arkime.com/wise). By default, Malcolm leverages the WISE plugin internally but does not expose the functionality to the end user:
        + `ARKIME_EXPOSE_WISE_GUI` - if set to `true` the WISE interface will be available at: `https://<MALCOLM-IP>/wise`. This defaults to `true`.
        + `ARKIME_ALLOW_WISE_GUI_CONFIG` - if set to `true` the WISE interface can be used to configure the WISE service. This only applies if `ARKIME_EXPOSE_WISE_GUI` is set to `true`. The default value is `true`.
        + `ARKIME_WISE_CONFIG_PIN_CODE` - the WISE service requires a configuration pin. This value will be required to save any WISE configuration changes.  The default value is `WISE2019`.
        + `ARKIME_WISE_SERVICE_URL` - to leverage WISE, arkime-capture needs to be provided a `wiseURL` value. The value of this environment variable is copied into the `wiseURL` value in arkime-live containers.
        + `WISE` - indicates if the WISE service is `on` or `off`. This environment variable defaults to `off`.
* **`arkime-live.env`** - settings for live traffic capture with Arkime
    - See [**Tuning Arkime**](live-analysis.md#LiveAnalysisTuningArkime) for variables related to managing Arkime's performance and resource utilization during live capture.
* **`auth-common.env`** - [authentication](authsetup.md)-related settings
    - `NGINX_AUTH_MODE` - valid values are `basic` (or `true` for legacy compatibility), use [TLS-encrypted HTTP basic](authsetup.md#AuthBasicAccountManagement) authentication (default); `ldap` (or `false` for legacy compatibility), use [Lightweight Directory Access Protocol (LDAP)](authsetup.md#AuthLDAP) authentication; `keycloak` to use [authentication managed by Malcolm's embedded Keycloak](authsetup.md#AuthKeycloakEmbedded) instance;  `keycloak_remote` to use [authentication managed by a remote Keycloak](authsetup.md#AuthKeycloakRemote) instance; `no_authentication` to disable authentication
    - `NGINX_REQUIRE_GROUP` and `NGINX_REQUIRE_ROLE` - When using [Keycloak authentication](authsetup.md#AuthKeycloak), setting these values will require authenticated users to [belong to groups and assigned roles](authsetup.md#AuthKeycloakReqGroupsRoles), respectively. Multiple values may be specified with a comma-separated list. Note that these requirements are cumulative: users must match all of the items specified. An empty value (default) means no group/role restriction is applied. [LDAP authentication](authsetup.md#AuthLDAP) can also require group membership, but that is specified in `nginx_ldap.conf` by setting `require group` rather than in `auth-common.env`.
    - `ROLE…` - variables used to manage [role-based access control](authsetup.md#AuthKeycloakRBAC)
* **`auth.env`** - stores the Malcolm administrator's username and password hash for its nginx reverse proxy
* **`beats-common.env`** - settings for interactions between [Logstash](https://www.elastic.co/products/logstash) and [Filebeat](https://www.elastic.co/products/beats/filebeat)
    - `BEATS_SSL` – if set to `true`, Logstash will use require encrypted communications for any external [Beats](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-beats.html)-based forwarders from which it will accept logs (default `true`)
    - `LOGSTASH_HOST` – the host and port at which Beats-based forwarders will connect to Logstash (default `logstash:5044`); see `MALCOLM_PROFILE` below
* **`dashboards.env`** and **`dashboards-helper.env`** - settings for the containers that configure and maintain [OpenSearch](https://opensearch.org/) and [OpenSearch Dashboards](https://opensearch.org/docs/latest/dashboards/index/)
    - `DASHBOARDS_URL` - used primarily when `OPENSEARCH_PRIMARY` is set to `elasticsearch-remote` (see [OpenSearch and Elasticsearch instances](opensearch-instances.md#OpenSearchInstance)), this variable stores the URL for the [Kibana](https://www.elastic.co/kibana) instance into which Malcolm's dashboard's and index templates will be imported
    - `DASHBOARDS_PREFIX` – a string to prepend to the titles of Malcolm's prebuilt [dashboards](dashboards.md#PrebuiltVisualizations) prior upon import during Malcolm's initialization (default is an empty string)
    - `DASHBOARDS_DARKMODE` – if set to `true`, [OpenSearch Dashboards](dashboards.md#DashboardsVisualizations) will be set to dark mode upon initialization (default `true`)
    - `DASHBOARDS_TIMEPICKER_FROM` and `DASHBOARDS_TIMEPICKER_TO` – sets the "from" and "to" values, respectively, for OpenSearch Dashboard's `timepicker:timeDefaults` [setting](https://docs.opensearch.org/latest/dashboards/management/advanced-settings/#general-settings) (default `now-24h` and `now`, meaning "last 24 hours")
    -  – if set to `true`, [OpenSearch Dashboards](dashboards.md#DashboardsVisualizations) will be set to dark mode upon initialization (default `true`)
    - `OPENSEARCH_INDEX_SIZE_PRUNE_LIMIT` - the maximum cumulative size of OpenSearch indices are allowed to consume before the oldest indices are deleted, see [**Managing disk usage**](#DiskUsage) below
* **`dashboards-secret.env`** - secret values used by the containers that configure and maintain OpenSearch and OpenSearch Dashboards
    - `MALCOLM_API_LOOPBACK_TOKEN` - a shared secret used to authenticate "internal" calls to the Malcolm API [event logging](api-event-logging.md) webhook (e.g., the OpenSearch Alerting loopback webhook); if left blank (the default) no internal bypass is allowed and the webhook is subject to the same authentication/authorization as other API requests
* **`filebeat.env`** - settings specific to [Filebeat](https://www.elastic.co/products/beats/filebeat), particularly for how Filebeat watches for new log files to parse and how it receives and stores [third-Party logs](third-party-logs.md)
    - `LOG_CLEANUP_MINUTES` and `ZIP_CLEANUP_MINUTES` - these variables deal cleaning up already-processed log files, see [**Managing disk usage**](#DiskUsage) below
    - The following variables configure Malcolm's ability to [accept syslog](https://www.elastic.co/guide/en/beats/filebeat/current/syslog.html) messages:
        + `FILEBEAT_SYSLOG_TCP_LISTEN` and `FILEBEAT_SYSLOG_UDP_LISTEN` - if set to `true`, Malcolm will accept syslog messages over TCP and/or UDP, respectively
        + `FILEBEAT_SYSLOG_TCP_PORT` and `FILEBEAT_SYSLOG_UDP_PORT` - the port on which Malcolm will accept syslog messages over TCP and/or UDP, respectively
            * If Malcolm is running in an instance installed via the [Malcolm installer ISO](malcolm-iso.md#ISO), please see also [ISO-installed Desktop Environment Firewall](third-party-logs.md#SyslogISOFirewall).
        + `FILEBEAT_SYSLOG_TCP_FORMAT` and `FILEBEAT_SYSLOG_UDP_FORMAT` - one of `auto`, `rfc3164`, or `rfc5424`, to specify the allowed format for syslog messages over TCP and/or UDP, respectively (default `auto`)
        + `FILEBEAT_SYSLOG_TCP_MAX_MESSAGE_SIZE` and `FILEBEAT_SYSLOG_UDP_MAX_MESSAGE_SIZE` - defines the maximum message size of the message received over TCP and/or UDP, respectively (default: `10KiB` for UDP, `20MiB` for TCP)
        + `FILEBEAT_SYSLOG_TCP_MAX_CONNECTIONS` - specifies the maximum current number of TCP connections for syslog messages
        + `FILEBEAT_SYSLOG_TCP_SSL` - if set to `true`, syslog messages over TCP will require the use of TLS. When [`./scripts/auth_setup`](authsetup.md#AuthSetup) is run, self-signed certificates are generated which may be used by remote log forwarders. Located in the `filebeat/certs/` directory, the certificate authority and client certificate and key files should be copied to the host on which the forwarder is running and used when defining its settings for connecting to Malcolm.
* **`filescan.env`**, **`filescan-secret.env`**, and **`pipeline.env`** - Settings related to [scanning of automatically-extracted files observed in traffic](file-scanning.md#ZeekFileExtraction) using [Strelka](https://target.github.io/strelka/#/) (see also **`zeek.env`** below)
    - `CLAMD_…` - variables used to managed the [ClamAV](https://www.clamav.net/) server to which Strelka can submit files for scanning
    - `FILESCAN_HTTP_SERVER_ENABLE` – if set to `true`, the directory containing [Zeek-extracted files](file-scanning.md#ZeekFileExtraction) will be served over HTTP at `./extracted-files/` (e.g., **https://localhost/extracted-files/** if connecting locally)
    - `FILESCAN_HTTP_SERVER_KEY` – specifies the password for the ZIP archive if `FILESCAN_HTTP_SERVER_ZIP` is `true`; otherwise, this specifies the decryption password for encrypted Zeek-extracted files in an `openssl enc`-compatible format (e.g., `openssl enc -aes-256-cbc -d -in example.exe.encrypted -out example.exe`)
    - `FILESCAN_HTTP_SERVER_ZIP` – if to `true`, the Zeek-extracted files will be archived in a ZIP file upon download
    - `FILESCAN_PRESERVATION` – determines behavior for preservation of [Zeek-extracted files](file-scanning.md#ZeekFileExtraction)
    - `FILESCAN_PRUNE_INTERVAL_SECONDS` - the interval between checking the prune conditions, in seconds (default `300`)
    - `FILESCAN_PRUNE_THRESHOLD_MAX_SIZE` - specifies the maximum size, specified either in gigabytes or as a human-readable data size (e.g., `250G`), that the  `./zeek-logs/extract_files/` directory is allowed to contain before the prune condition triggers
    - `FILESCAN_PRUNE_THRESHOLD_TOTAL_DISK_USAGE_PERCENT` - specifies a maximum fill percentage for the file system containing the `./zeek-logs/extract_files/`; in other words, if the disk is more than this percentage utilized, the prune condition triggers
    - `PIPELINE_DISABLED` - if set to `true`, file scanning with Strelka will be disabled
    - `RULES_UPDATE_ENABLED` – if set to `true`, file scanner engines (e.g., ClamAV, YARA, etc.) will periodically update their rule definitions (default `false`)
    - `STRELKA_BACKEND_PROCS` - specifies the number of Strelka backend instances performing file scanning concurrently (default `1`)
    - `STRELKA_SCANNERS` - comma-separated list of [Strelka scanners](https://target.github.io/strelka/#/?id=scanner-list) to enable
    - `YARA_CUSTOM_RULES_ONLY` - if set to `true`, Malcolm will bypass its default YARA rulesets and use only [user-defined rules](custom-rules.md#YARA) in `./yara/rules` (default `false`)
* **`keycloak.env`** - settings specific to [Keycloak](https://www.keycloak.org/)
    - The following variables are used for all [Keycloak](authsetup.md#AuthKeycloak) configurations, be it Malcolm's [embedded instance](authsetup.md#AuthKeycloakEmbedded) or a [remote instance](authsetup.md#AuthKeycloakRemote) (see `NGINX_AUTH_MODE` above):
        + `KEYCLOAK_AUTH_REALM` - specifies the name of the Keycloak [realm](https://www.keycloak.org/docs/latest/server_admin/index.html#_configuring-realms) (default [`master`](https://www.keycloak.org/docs/latest/server_admin/index.html#the-master-realm))
        + `KEYCLOAK_AUTH_REDIRECT_URI` - specifies the relative path which is the Malcolm URI to which Keycloak will redirect users after a successful authentication (default `/index.html`, which will redirect users to the Malcolm landing page)
        + `KEYCLOAK_AUTH_URL` - specifies the Keycloak endpoint URL, or the URL to which Malcolm should direct authentication requests for Keycloak. If a [remote Keycloak](authsetup.md#AuthKeycloakRemote) instance is being used, this would be the URL for that instance (e.g., **https://keycloak.example.com** ). If Malcolm is using its embedded [Keycloak instance](authsetup.md#AuthKeycloakEmbedded), this host portion of the URL should be the hostname or IP address at which Malcolm is available, followed by **/keycloak** (or whatever the value of `KC_HTTP_RELATIVE_PATH` has been set to; see below) (e.g., **https://malcolm.internal.lan/keycloak** or **https://192.168.100.10/keycloak** ).
        + `KEYCLOAK_CLIENT_ID` and `KEYCLOAK_CLIENT_SECRET` - identify the Keycloak client Malcolm will use and the secret associated with that client
    - The following variables are only used for Malcolm's embedded [Keycloak instance](authsetup.md#AuthKeycloakEmbedded) (see [**All configuration**](https://www.keycloak.org/server/all-config) in the Keycloak guide for more details on these values):
        + [`KC_CACHE`](https://www.keycloak.org/server/all-config#category-cache) - defines the cache mechanism for high-availability (default `local` as Malcolm's embedded Keycloak instance is single-node)
        + [`KC_HEALTH_ENABLED`](https://www.keycloak.org/server/all-config#category-health) - if set to `true`, enables the health check endpoint used internally by the container health script
        + [`KC_HOSTNAME`](https://www.keycloak.org/server/all-config#category-hostname_v2) - address at which the Keycloak server is exposed (defaults to blank, as `KC_HOSTNAME_STRICT` below defaults to `false`)
        + [`KC_HOSTNAME_STRICT`](https://www.keycloak.org/server/all-config#category-hostname_v2) - if set to `true`, disables dynamically resolving the hostname from request headers
        + [`KC_HTTP_ENABLED`](https://www.keycloak.org/server/all-config#category-http) - enables the HTTP listener (default `true` as Malcolm is [proxying](https://www.keycloak.org/server/reverseproxy) the embedded Keycloak instance behind nginx)
        + [`KC_HTTP_RELATIVE_PATH`](https://www.keycloak.org/server/all-config#category-http) - specifies the Malcolm path under which Keycloak serves resources (should not be changed from its default value of `/keycloak`)
        + [`KC_METRICS_ENABLED`](https://www.keycloak.org/server/all-config#category-metrics) - specifies if the server should expose metrics (default `false`)
        + [`KC_PROXY_HEADERS`](https://www.keycloak.org/server/all-config#category-proxy) - the proxy headers that should be accepted by Keycloak (should not be changed from its default value of `xforwarded`)
        + `KC_BOOTSTRAP_ADMIN_USERNAME` and `KC_BOOTSTRAP_ADMIN_PASSWORD` - values for boostrapping the temporary Keycloak admin service account (see [Keycloak configuration](authsetup.md#AuthKeycloakEmbedded))
* **`logstash.env`** - settings specific to [Logstash](https://www.elastic.co/products/logstash)
    - `LOGSTASH_OUI_LOOKUP` – if set to `true`, Logstash will map MAC addresses to vendors for all source and destination MAC addresses when analyzing Zeek logs (default `true`)
    - `LOGSTASH_REVERSE_DNS` – if set to `true`, Logstash will perform a reverse DNS lookup for all external source and destination IP address values when analyzing Zeek logs (default `false`)
    - `LOGSTASH_SEVERITY_SCORING` - if set to `true`, Logstash will perform [severity scoring](severity.md#Severity) when analyzing Zeek logs (default `true`)
    - `LS_JAVA_OPTS` - part of LogStash's [JVM settings](https://www.elastic.co/guide/en/logstash/current/jvm-settings.html), the `-Xmx` and `-Xms` values set the size of LogStash's Java heap (we recommend somewhere between `1500m` and `4g`)
    * `pipeline.workers`, `pipeline.batch.size` and `pipeline.batch.delay` - these settings are used to tune the performance and resource utilization of the the `logstash` container; see [Tuning and Profiling Logstash Performance](https://www.elastic.co/guide/en/logstash/current/tuning-logstash.html), [`logstash.yml`](https://www.elastic.co/guide/en/logstash/current/logstash-settings-file.html) and [Multiple Pipelines](https://www.elastic.co/guide/en/logstash/current/multiple-pipelines.html)
* **`lookup-common.env`** - settings for enrichment lookups, including those used for [customizing event severity scoring](severity.md#SeverityConfig)
    - `FREQ_LOOKUP` - if set to `true`, domain names (from DNS queries and SSL server names) will be assigned entropy scores as calculated by [`freq`](https://github.com/MarkBaggett/freq) (default `false`)
    - `FREQ_SEVERITY_THRESHOLD` - when [severity scoring](severity.md#Severity) is enabled, this variable indicates the entropy threshold for assigning severity to events with entropy scores calculated by [`freq`](https://github.com/MarkBaggett/freq); a lower value will only assign severity scores to fewer domain names with higher entropy (e.g., `2.0` for `NQZHTFHRMYMTVBQJE.COM`), while a higher value will assign severity scores to more domain names with lower entropy (e.g., `7.5` for `naturallanguagedomain.example.org`) (default `2.0`)
    - `SENSITIVE_COUNTRY_CODES` - when [severity scoring](severity.md#Severity) is enabled, this variable defines a comma-separated list of sensitive countries (using [ISO 3166-1 alpha-2 codes](https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Current_codes)) (default `'AM,AZ,BY,CN,CU,DZ,GE,HK,IL,IN,IQ,IR,KG,KP,KZ,LY,MD,MO,PK,RU,SD,SS,SY,TJ,TM,TW,UA,UZ'`, taken from the U.S. Department of Energy Sensitive Country List)
    - `TOTAL_MEGABYTES_SEVERITY_THRESHOLD` - when [severity scoring](severity.md#Severity) is enabled, this variable indicates the size threshold (in megabytes) for assigning severity to large connections or file transfers (default `1000`)
* **`netbox-common.env`**, `netbox.env` and `netbox-secret.env` - settings related to [NetBox](https://netbox.dev/) and [Asset Interaction Analysis](asset-interaction-analysis.md#AssetInteractionAnalysis)
    - `NETBOX_MODE` - determine whether Malcolm will start and manage a [NetBox](asset-interaction-analysis.md#AssetInteractionAnalysis) instance; valid values are `local` (use an embedded instance NetBox), `remote` (use a remote instance of NetBox), or `disabled` (the default)
    - `NETBOX_ENRICHMENT` - if set to `true`, Logstash will [enrich network traffic metadata](asset-interaction-analysis.md#NetBoxEnrichment) via NetBox API calls
    - `NETBOX_DEFAULT_SITE` - specifies the default NetBox [site name](https://demo.netbox.dev/static/docs/core-functionality/sites-and-racks/) for use when [enriching network traffic metadata via NetBox lookups](asset-interaction-analysis.md#NetBoxEnrichment) if a specific site is not otherwise specified for the source of the data (default `Malcolm`)
    - `NETBOX_AUTO_POPULATE` - if set to `true`, Logstash will [populate the NetBox inventory](asset-interaction-analysis.md#NetBoxPopPassive) based on observed network traffic
    - `NETBOX_AUTO_POPULATE_SUBNETS` - a comma-separated list of private CIDR subnets to control NetBox IP autopopulation (see [**Subnets considered for autopopulation**](asset-interaction-analysis.md#NetBoxAutoPopSubnets); default is an empty string, meaning all private IPv4 and IPv6 ranges are autopopulated)
    - `NETBOX_AUTO_CREATE_PREFIX` - if set to `true`, Logstash will automatically create private subnet prefixes in the [NetBox inventory](asset-interaction-analysis.md#NetBoxPopPassive) based on observed network traffic
    - `NETBOX_DEFAULT_AUTOCREATE_MANUFACTURER` - if set to `true`, new manufacturer entries will be created in the NetBox database when [matching device manufacturers to OUIs](asset-interaction-analysis.md#NetBoxPopPassiveOUIMatch) (default `true`)
    - `NETBOX_DEFAULT_FUZZY_THRESHOLD` - fuzzy-matching threshold for [matching device manufacturers to OUIs](asset-interaction-analysis.md#NetBoxPopPassiveOUIMatch) (default `0.95`)
    - The following variables should only be set if `NETBOX_MODE` is set to `remote`, otherwise they should be blank:
        + `NETBOX_URL` - the URL of the remote NetBox instance (e.g., `https://netbox.example.org` or `https://example.com/netbox`)
        + `NETBOX_TOKEN` - the [API token](https://netboxlabs.com/docs/netbox/en/stable/integrations/rest-api/#tokens) for the remote NetBox instance (40 hexadecimal characters)
* **`nginx.env`** - settings specific to Malcolm's nginx reverse proxy
    - `NGINX_LOG_ACCESS_AND_ERRORS` - if set to `true`, all access to Malcolm via its [web interfaces](quickstart.md#UserInterfaceURLs) will be logged to OpenSearch (default `false`)
    - `NGINX_SSL` - if set to `true`, require HTTPS connections to Malcolm's `nginx-proxy` container (default); if set to `false`, use unencrypted HTTP connections (using unsecured HTTP connections is **NOT** recommended unless you are running Malcolm behind another reverse proxy such as Traefik, Caddy, etc.) Also note: in some circumstances disabling SSL in NGINX while leaving SSL enabled in Arkime can result in a "Missing token" Arkime error. This is due to Arkime's Cross-Site Request Forgery mitigation cookie being passed to the browser with the "secure" flag enabled.
    - `NGINX_X_FORWARDED_PROTO_OVERRIDE`
    - The following variables control nginx's [resolver directive](https://nginx.org/en/docs/http/ngx_http_core_module.html#resolver). Note that these settings do not affect Malcolm's ability to capture or inspect IPv4/IPv6 traffic: they are only used if and when nginx itself needs to resolve hostnames in the network in which Malcolm resides.
        + `NGINX_RESOLVER_OVERRIDE` - if set, overrides automatic detection of the resolver address used (default is unset)
        + `NGINX_RESOLVER_IPV4` - if `false`, sets the `ipv4=off` parameter in the resolver directive (default is `true`)
        + `NGINX_RESOLVER_IPV6` - if `false`, sets the `ipv6=off` parameter in the resolver directive; it is recommended to set this to `false` if your network does not support IPv6 (default is `true`)
* **`opensearch.env`** - settings specific to [OpenSearch](https://opensearch.org/)
    - `OPENSEARCH_JAVA_OPTS` - one of OpenSearch's most [important settings](https://opensearch.org/docs/latest/install-and-configure/install-opensearch/index/#important-settings), the `-Xmx` and `-Xms` values set the size of OpenSearch's Java heap (we recommend setting this value to half of system RAM, up to 32 gigabytes)
    - `OPENSEARCH_PRIMARY` - one of `opensearch-local`, `opensearch-remote`, or `elasticsearch-remote`, to determine the [OpenSearch or Elasticsearch instance](opensearch-instances.md#OpenSearchInstance) Malcolm will use  (default `opensearch-local`)
    - `OPENSEARCH_URL` - when using Malcolm's internal OpenSearch instance (i.e., `OPENSEARCH_PRIMARY` is `opensearch-local`) this should be `https://opensearch:9200`, otherwise this value specifies the primary remote instance URL in the format `protocol://host:port` (default `https://opensearch:9200`)
    - `OPENSEARCH_SSL_CERTIFICATE_VERIFICATION` - if set to `true`, connections to the primary remote OpenSearch instance will require full TLS certificate validation (this may fail if using self-signed certificates) (default `false`)
    - `OPENSEARCH_SECONDARY` - one of `opensearch-local`, `opensearch-remote`, `elasticsearch-remote`, or blank (unset) to indicate that Malcolm should forward logs to a secondary remote OpenSearch instance in addition to the primary OpenSearch instance (default is unset)
    - `OPENSEARCH_SECONDARY_URL` - when forwarding to a secondary remote OpenSearch instance (i.e., `OPENSEARCH_SECONDARY` is set) this value specifies the secondary remote instance URL in the format `protocol://host:port`

api/project/config.py
import os


basedir = os.path.abspath(os.path.dirname(__file__))


class Config(object):
    ARKIME_FIELDS_INDEX = os.getenv('ARKIME_FIELDS_INDEX', 'arkime_fields')
    ARKIME_HOST = os.getenv('ARKIME_HOST', 'arkime')
    ARKIME_NETWORK_INDEX_PATTERN = os.getenv('ARKIME_NETWORK_INDEX_PATTERN', 'arkime_sessions3-*')
    ARKIME_NETWORK_INDEX_TIME_FIELD = os.getenv('ARKIME_NETWORK_INDEX_TIME_FIELD', 'firstPacket')
    ARKIME_PORT = int(os.getenv('ARKIME_VIEWER_PORT', os.getenv('ARKIME_PORT', '8005')).split(':')[-1])
    ARKIME_SSL = os.getenv('ARKIME_SSL', 'true')
    ARKIME_STATS_INDEX_PATTERN = "arkime_stats_*"
    BUILD_DATE = os.getenv('BUILD_DATE', 'unknown')
    DASHBOARDS_HELPER_HOST = os.getenv('DASHBOARDS_HELPER_HOST', 'dashboards-helper')
    DASHBOARDS_MAPS_PORT = int(os.getenv('DASHBOARDS_MAPS_PORT', '28991'))
    DASHBOARDS_URL = os.getenv('DASHBOARDS_URL', 'http://dashboards:5601/dashboards')
    DOCTYPE_DEFAULT = os.getenv('DOCTYPE_DEFAULT', 'network')
    FILEBEAT_HOST = os.getenv('FILEBEAT_HOST', 'filebeat')
    FILEBEAT_TCP_JSON_PORT = int(os.getenv('FILEBEAT_TCP_JSON_PORT', '5045'))
    FILESCAN_HOST = os.getenv('FILESCAN_HOST', 'filescan')
    FILESCAN_HEALTH_PORT = int(os.getenv('FILESCAN_HEALTH_PORT', '8001'))
    FILESCAN_HTTP_SERVER_PORT = int(os.getenv('FILESCAN_HTTP_SERVER_PORT', '8006'))
    FREQ_URL = os.getenv('FREQ_URL', 'http://freq:10004')
    LOGSTASH_API_PORT = int(os.getenv('LOGSTASH_API_PORT', '9600'))
    LOGSTASH_HOST = os.getenv('LOGSTASH_HOST', 'logstash')
    LOGSTASH_LJ_PORT = int(os.getenv('LOGSTASH_LJ_PORT', '5044'))
    MALCOLM_API_DEBUG = os.getenv('MALCOLM_API_DEBUG', 'false')
    MALCOLM_API_LOOPBACK_TOKEN = os.getenv('MALCOLM_API_LOOPBACK_TOKEN', '')
    MALCOLM_API_PREFIX = os.getenv('MALCOLM_API_PREFIX', 'mapi')
    MALCOLM_NETWORK_INDEX_PATTERN = os.getenv('MALCOLM_NETWORK_INDEX_PATTERN', 'arkime_sessions3-*')
    MALCOLM_NETWORK_INDEX_TIME_FIELD = os.getenv('MALCOLM_NETWORK_INDEX_TIME_FIELD', 'firstPacket')
    MALCOLM_OTHER_INDEX_PATTERN = os.getenv('MALCOLM_OTHER_INDEX_PATTERN', 'malcolm_beats_*')
    MALCOLM_OTHER_INDEX_TIME_FIELD = os.getenv('MALCOLM_OTHER_INDEX_TIME_FIELD', '@timestamp')
    MALCOLM_TEMPLATE = os.getenv('MALCOLM_TEMPLATE', 'malcolm_template')
    MALCOLM_VERSION = os.getenv('MALCOLM_VERSION', 'unknown')
    NETBOX_TOKEN = os.getenv('NETBOX_TOKEN') or os.getenv('SUPERUSER_API_TOKEN', '')
    NETBOX_URL = os.getenv('NETBOX_URL') or 'http://netbox:8080/netbox'
    OPENSEARCH_CREDS_CONFIG_FILE = os.getenv(
        'OPENSEARCH_CREDS_CONFIG_FILE', '/var/local/curlrc/.opensearch.primary.curlrc'
    )
    OPENSEARCH_PRIMARY = os.getenv('OPENSEARCH_PRIMARY', 'opensearch-local')
    OPENSEARCH_SSL_CERTIFICATE_VERIFICATION = os.getenv('OPENSEARCH_SSL_CERTIFICATE_VERIFICATION', 'false')
    OPENSEARCH_URL = os.getenv('OPENSEARCH_URL', 'https://opensearch:9200')
    PCAP_MONITOR_HOST = os.getenv('PCAP_MONITOR_HOST', 'pcap-monitor')
    PCAP_TOPIC_PORT = int(os.getenv('PCAP_TOPIC_PORT', '30441'))
    VALKEY_CACHE_HOST = os.getenv('VALKEY_CACHE_HOST', 'valkey')
    VALKEY_CACHE_PORT = int(os.getenv('VALKEY_CACHE_PORT', '6379'))
    VALKEY_HOST = os.getenv('VALKEY_HOST', 'valkey')
    VALKEY_PORT = int(os.getenv('VALKEY_PORT', '6379'))
    VALKEY_PASSWORD = os.getenv('VALKEY_PASSWORD', None)
    RESULT_SET_LIMIT = int(os.getenv('RESULT_SET_LIMIT', '500'))
    ROLE_ADMIN = os.getenv('ROLE_ADMIN', '')
    ROLE_BASED_ACCESS = os.getenv('ROLE_BASED_ACCESS', 'false')
    ROLE_DASHBOARDS_READ_ACCESS = os.getenv('ROLE_DASHBOARDS_READ_ACCESS', '')
    ROLE_DASHBOARDS_READ_ALL_APPS_ACCESS = os.getenv('ROLE_DASHBOARDS_READ_ALL_APPS_ACCESS', '')
    ROLE_DASHBOARDS_READ_WRITE_ACCESS = os.getenv('ROLE_DASHBOARDS_READ_WRITE_ACCESS', '')
    ROLE_DASHBOARDS_READ_WRITE_ALL_APPS_ACCESS = os.getenv('ROLE_DASHBOARDS_READ_WRITE_ALL_APPS_ACCESS', '')
    ROLE_NETBOX_READ_ACCESS = os.getenv('ROLE_NETBOX_READ_ACCESS', '')
    ROLE_NETBOX_READ_WRITE_ACCESS = os.getenv('ROLE_NETBOX_READ_WRITE_ACCESS', '')
    ROLE_READ_ACCESS = os.getenv('ROLE_READ_ACCESS', '')
    ROLE_READ_WRITE_ACCESS = os.getenv('ROLE_READ_WRITE_ACCESS', '')
    ROLE_UPLOAD = os.getenv('ROLE_UPLOAD', '')
    STRELKA_HOST = os.getenv('STRELKA_HOST', 'strelka-frontend')
    STRELKA_PORT = int(os.getenv('STRELKA_PORT', '57314'))
    VCS_REVISION = os.getenv('VCS_REVISION', 'unknown')
